experimental/components/VideoPlayer/library/src/desktopMain/kotlin/org/jetbrains/compose/videoplayer/DesktopVideoPlayer.kt

```diff
@@ -1,63 +1,137 @@
+internal actual fun VideoPlayerImpl(
+    url: String,
+    isResumed: Boolean,
+    volume: Float,
+    speed: Float,
+    seek: Float,
+    isFullscreen: Boolean,
+    modifier: Modifier,
+    onFinish: (() -> Unit)?
+): State<Progress> {
+    val mediaPlayerComponent = initializeMediaPlayerComponent()
+    val mediaPlayer = remember { mediaPlayerComponent.mediaPlayer() }
+    mediaPlayer.setupVideoFinishHandler(onFinish)
+
+    val factory = remember { { mediaPlayerComponent } }
+    /* OR the following code and using SwingPanel(factory = { factory }, ...) */
+    // val factory by rememberUpdatedState(mediaPlayerComponent)
+
+    LaunchedEffect(url) { mediaPlayer.media().play(url) /* OR .start(url) */ }
+    LaunchedEffect(seek) { mediaPlayer.controls().setPosition(seek) }
+    LaunchedEffect(speed) { mediaPlayer.controls().setRate(speed) }
+    LaunchedEffect(volume) { mediaPlayer.audio().setVolume(volume.toPercentage()) }
+    LaunchedEffect(isResumed) { mediaPlayer.controls().setPause(!isResumed) }
+    LaunchedEffect(isFullscreen) {
+        if (mediaPlayer is EmbeddedMediaPlayer) {
+            /*
+             * To be able to access window in the commented code below,
+             * extend the player composable function from WindowScope.
+             * See https://github.com/JetBrains/compose-jb/issues/176#issuecomment-812514936
+             * and its subsequent comments.
+             *
+             * We could also just fullscreen the whole window:
+             * `window.placement = WindowPlacement.Fullscreen`
+             * See https://github.com/JetBrains/compose-multiplatform/issues/1489
+             */
+            // mediaPlayer.fullScreen().strategy(ExclusiveModeFullScreenStrategy(window))
+            mediaPlayer.fullScreen().toggle()
         }
     }
-    SideEffect {
-        mediaPlayerComponent.mediaPlayer().media().play(url)
+    DisposableEffect(Unit) { onDispose { mediaPlayer.release() } }
+    SwingPanel(
+        factory = factory,
+        background = Color.Transparent,
+        modifier = modifier
+    )
+    return mediaPlayer.produceProgress()
+}
+
+private fun Float.toPercentage() = (this * 100).roundToInt()
+
+/**
+ * See https://github.com/caprica/vlcj/issues/887#issuecomment-503288294
+ * for why we're using CallbackMediaPlayerComponent for macOS.
+ */
+@Composable
+private fun initializeMediaPlayerComponent(): Component = remember {
+    NativeDiscovery().discover()
+    if (isMacOS()) {
+        CallbackMediaPlayerComponent()
+    } else {
+        EmbeddedMediaPlayerComponent()
     }
+}
 
-    DisposableEffect(Unit) {
-        onDispose {
-            mediaPlayerComponent.mediaPlayer().release()
+/**
+ * We play the video on finish (so the player is kind of idempotent),
+ * unless the [onFinish] callback stops the playback.
+ * Using `mediaPlayer.controls().repeat = true` did not work as expected.
+ */
+@Composable
+private fun MediaPlayer.setupVideoFinishHandler(onFinish: (() -> Unit)?) =
```

Comment:
Usually, when I see a function declared with "=", I first think that it should return something, given that there is no explicit return type.
In this case it's Unit. So maybe it's better to add the body into "{ ... }" instead of "="

Reply:
Updated the code.